% SCGEATOOLBOX
%
% Files
%   cdgea                     - Changes to scGEAToolbox directory
%   demo_Getting_Started      - 
%   demo_script1              - Demonstration of Filter, Normalization and Batch Correction of Data in scGEAToolbox
%   demo_script2              - Demonstration of Feature Selection Functions in scGEAToolbox
%   demo_script3              - Demonstration of Visualization Functions in scGEAToolbox
%   demo_script4              - Demonstration of Clustering Functions in scGEAToolbox
%   demo_script5              - Demonstration of Pseudotime Analysis and Gene Network Functions in scGEAToolbox
%   demo_script6              - Demonstration of DE Analysis and Marker Gene Identification Functions in scGEAToolbox
%   e_filtadjc                - A - adjacency matrix
%   fun_cmp_clusters          - https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby076/5077112
%   fun_num_clusters          - 
%   i_plot_pseudotimeseries   - 
%   i_pseudotime_by_princurve - adding pseudotime trajectory curve using princurve
%   i_pseudotime_by_splinefit - adding pseudotime trajectory curve using splinefit
%   infox                     - 
%   installx                  - 
%   norm_deseq                - For DESeq normalization, the geometric mean for each gene was computed 
%   norm_libsize              - 
%   sc_cellcyclescoring       - https://rdrr.io/cran/Seurat/src/R/utilities.R
%   sc_celltypecaller         - (X,genelist,clusterid,varargin)
%   sc_cluster_s              - if min(size(s))>3, error('S is coordinates of dimensional reduction.'); end
%   sc_cluster_x              - 
%   sc_deg                    - https://satijalab.org/seurat/v3.1/de_vignette.html
%   sc_filterc                - 
%   sc_filterg                - 
%   sc_genestat               - 
%   sc_grn                    - construct single-cell gene regulatory network (scGRN)
%   sc_grnview                - 
%   sc_grnview2               - 
%   sc_heg                    - HEGs - highly expressed genes
%   sc_hvg                    - HVGs selection - This method uses the CV^2 on normalized count data to 
%   sc_knngraph               - Generate KNN group network from cell embeddings
%   sc_mergedata              - 
%   sc_mergesces              - 
%   sc_mtratio                - 
%   sc_norm                   - 
%   sc_pcnet                  - [A]=sc_pcnet(X,ncom)      % X = expression matrix of genes x cells
%   sc_pcnetdenoised          - A=sc_pcnetdenoised(X);
%   sc_pcnetpar               - [A]=sc_pcnetpar(X,ncom)
%   sc_phate                  - UMAP embedding of cells
%   sc_pickmarkers            - 
%   sc_potency                - Estimate differentiation potency of cells
%   sc_qcfilter               - 
%   sc_qcmetrics              - QC metrics
%   sc_read10xdir             - Read files from a 10x Genomics cellranger output folder
%   sc_readfile               - 
%   sc_readhdf5file           - https://www.mathworks.com/help/matlab/hdf5-files.html
%   sc_readmtxfile            - 
%   sc_readtsvfile            - 
%   sc_rmdugenes              - remove genes with duplicate name
%   sc_rmmtcells              - 
%   sc_rmmtgenes              - 
%   sc_scatter                - SC_SCATTER
%   sc_scatter_sce            - 
%   sc_scattermarker          - SC_SCATTERMARKER(X,genelist,g,s,methodid)
%   sc_sce2hdf5               - import h5py
%   sc_sce2rds                - 
%   sc_selectc                - 
%   sc_selectg                - Select genes by expression levels
%   sc_snndpc                 - Clustering cell embeddings using SNNDPC - a SNN clustering algorithm
%   sc_snngraph               - Generates SNN graph from embedding
%   sc_splinefit              - identify genes with a profile deviated from normal
%   sc_splinefit2             - Compare genes' expression profiles with spline fit regression
%   sc_stem3                  - 
%   sc_tenifoldnet            - 
%   sc_trajectory             - 
%   sc_transform              - 
%   sc_tsne                   - tSNE embedding of cells
%   sc_umap                   - UMAP embedding of cells
%   sc_veg                    - 
%   sc_writefile              - 
