function [txt]=gui_showrefinfo(reftag)

txt=[];
switch reftag
    case 'Geometric Sketching [PMID:31176620]'
        txt=['"Large-scale single-cell RNA sequencing (scRNA-seq) ' ...
                'studies that profile hundreds of thousands of cells are ' ...
                'becoming increasingly common, overwhelming existing ' ...
                'analysis pipelines. Here, we describe how to enhance ' ...
                'and accelerate single-cell data analysis by summarizing ' ...
                'the transcriptomic heterogeneity within a dataset using ' ...
                'a small subset of cells, which we refer to as a geometric' ...
                ' sketch. Our sketches provide more comprehensive ' ...
                'visualization of transcriptional diversity, capture rare' ...
                ' cell types with high sensitivity, and reveal biological' ...
                ' cell types via clustering."'];
    case 'MELD [PMID:33558698]'
        txt=['"The goal of MELD is to identify populations of cells that' ...
            ' are most affected by an experimental perturbation. Rather ' ...
            'than clustering the data first and calculating differential ' ...
            'abundance of samples within clusters, MELD provides a density' ...
            ' estimate for each scRNA-seq sample for every cell in each ' ...
            'dataset. Comparing the ratio between the density of each ' ...
            'sample provides a quantitative estimate the effect of a ' ...
            'perturbation at the single-cell level. We can then identify ' ...
            'the cells most or least affected by the perturbation."'];
    case 'DecontX [PMID:32138770]'
        txt=['"Droplet-based microfluidic devices have become widely ' ...
            'used to perform single-cell RNA sequencing (scRNA-seq). ' ...
            'However, ambient RNA present in the cell suspension can ' ...
            'be aberrantly counted along with a cellâ€™s native mRNA ' ...
            'and result in cross-contamination of transcripts between ' ...
            'different cell populations. DecontX is a novel Bayesian ' ...
            'method to estimate and remove contamination in individual ' ...
            'cells. DecontX accurately predicts contamination levels in ' ...
            'a mouse-human mixture dataset and removes aberrant ' ...
            'expression of marker genes in PBMC datasets. We also ' ...
            'compare the contamination levels between four different ' ...
            'scRNA-seq protocols. Overall, DecontX can be incorporated ' ...
            'into scRNA-seq workflows to improve downstream analyses."'];
    case 'Harmony [PMID:31740819]'
        txt=['"The emerging diversity of single-cell RNA-seq datasets ' ...
            'allows for the full transcriptional characterization of cell ' ...
            'types across a wide variety of biological and clinical ' ...
            'conditions. However, it is challenging to analyze them ' ...
            'together, particularly when datasets are assayed with ' ...
            'different technologies, because biological and technical ' ...
            'differences are interspersed. We present Harmony, an ' ...
            'algorithm that projects cells into a shared embedding ' ...
            'in which cells group by cell type rather than dataset-' ...
            'specific conditions. Harmony simultaneously accounts for ' ...
            'multiple experimental and biological factors."'];
    case 'Scrublet [PMID:30954476]'
        txt=['ScRNA-seq experiments generate ''multiplet'' errors when ' ...
            ' multiple cells are labeled with the same barcode. ' ...
            'Wolock et al. describe Scrublet, a method for predicting ' ...
            'the effects of multiplets on downstream analyses and ' ...
            'identifying problematic multiplets.'];
    case 'scTenifoldNet [PMID:33336197]'
        txt=['"We present scTenifoldNet-a machine learning workflow built ' ...
            'upon principal-component regression, low-rank tensor ' ...
            'approximation, and manifold alignment-for constructing ' ...
            'and comparing single-cell gene regulatory networks ' ...
            '(scGRNs) using data from single-cell RNA sequencing. ' ...
            'scTenifoldNet reveals regulatory changes in gene ' ...
            'expression between samples by comparing the constructed ' ...
            'scGRNs. With real data, scTenifoldNet identifies ' ...
            'specific gene expression programs associated with ' ...
            'different biological processes, providing critical ' ...
            'insights into the underlying mechanism of regulatory ' ...
            'networks governing cellular transcriptional activities."'];
     case 'scTenifoldKnk [PMID:35510185]'
        txt=['"Gene knockout (KO) experiments are a proven, powerful ' ...
            'approach for studying gene function. However, systematic ' ...
            'KO experiments targeting a large number of genes are ' ...
            'usually prohibitive due to the limit of experimental ' ...
            'and animal resources. Here, we present scTenifoldKnk, ' ...
            'an efficient virtual KO tool that enables systematic KO ' ...
            'investigation of gene function using data from single-' ...
            'cell RNA sequencing (scRNA-seq). In scTenifoldKnk analysis, ' ...
            'a gene regulatory network (GRN) is first constructed ' ...
            'from scRNA-seq data of wild-type samples, and a target ' ...
            'gene is then virtually deleted from the constructed GRN. ' ...
            'Manifold alignment is used to align the resulting reduced ' ...
            'GRN to the original GRN to identify differentially ' ...
            'regulated genes, which are used to infer target gene ' ...
            'functions in analyzed cells. We demonstrate that the ' ...
            'scTenifoldKnk-based virtual KO analysis recapitulates ' ...
            'the main findings of real-animal KO experiments and ' ...
            'recovers the expected functions of genes in relevant ' ...
            'cell types."'];
    case 'GenKI [PMID:37246643]'
        txt=['"In this paper, we introduce Gene Knockout Inference ' ...
            '(GenKI), a virtual knockout (KO) tool for gene function ' ...
            'prediction using single-cell RNA sequencing (scRNA-seq) ' ...
            'data in the absence of KO samples when only wild-type (WT) ' ...
            'samples are available. Without using any information from ' ...
            'real KO samples, GenKI is designed to capture shifting ' ...
            'patterns in gene regulation caused by the KO perturbation ' ...
            'in an unsupervised manner and provide a robust and ' ...
            'scalable framework for gene function studies. To achieve ' ...
            'this goal, GenKI adapts a variational graph autoencoder ' ...
            '(VGAE) model to learn latent representations of genes ' ...
            'and interactions between genes from the input WT scRNA-seq ' ...
            'data and a derived single-cell gene regulatory network ' ...
            '(scGRN). The virtual KO data is then generated by ' ...
            'computationally removing all edges of the KO gene-the' ...
            ' gene to be knocked out for functional study-from the ' ...
            'scGRN. The differences between WT and virtual KO data are' ...
            ' discerned by using their corresponding latent parameters' ...
            ' derived from the trained VGAE model. Our simulations ' ...
            'show that GenKI accurately approximates the perturbation ' ...
            'profiles upon gene KO and outperforms the state-of-the-art ' ...
            'under a series of evaluation conditions. Using publicly ' ...
            'available scRNA-seq data sets, we demonstrate that ' ...
            'GenKI recapitulates discoveries of real-animal KO ' ...
            'experiments and accurately predicts cell type-specific ' ...
            'functions of KO genes. Thus, GenKI provides an in-silico ' ...
            'alternative to KO experiments that may partially replace ' ...
            'the need for genetically modified animals or other' ...
            ' genetically perturbed systems."'];
    case 'scTenifoldXct [PMID:36787742]'
        txt=['"We present scTenifoldXct, a semi-supervised computational ' ...
            'tool for detecting ligand-receptor (LR)-mediated cell-cell ' ...
            'interactions and mapping cellular communication graphs. ' ...
            'Our method is based on manifold alignment, using LR pairs ' ...
            'as inter-data correspondences to embed ligand and receptor' ...
            ' genes expressed in interacting cells into a unified latent' ...
            ' space. Neural networks are employed to minimize the ' ...
            'distance between corresponding genes while preserving ' ...
            'the structure of gene regression networks. We apply ' ...
            'scTenifoldXct to real datasets for testing and ' ...
            'demonstrate that our method detects interactions with' ...
            ' high consistency compared with other methods. ' ...
            'More importantly, scTenifoldXct uncovers weak but ' ...
            'biologically relevant interactions overlooked by' ...
            ' other methods. We also demonstrate how scTenifoldXct' ...
            ' can be used to compare different samples, such as ' ...
            'healthy vs. diseased and wild type vs. knockout, to ' ...
            'identify differential interactions, thereby revealing ' ...
            'functional implications associated with changes in ' ...
            'cellular communication status."'];
    otherwise
        return;
end
if ~isempty(txt)
    uiwait(helpdlg(txt,reftag));
end